### PR TITLE
Ray projects schema and validation

### PR DESCRIPTION
**Pull request by pcmoritz**
## What do these changes do?

This adds a schema and validation code for the ray project format.

## Linter

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/projects/projects.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    Args:
+        dir (str): Directory to start the search in.
+
+    Returns:
+        Path of the parent directory containing the .rayproject or
+        None if no such project is found.
+    """
+    prev, dir = None, os.path.abspath(dir)
+    while prev != dir:
+        if os.path.isdir(os.path.join(dir, ".rayproject")):
```
Comment (simon-mo) [inline python/ray/projects/projects.py]: Should it be suffix by `.json` or `.yaml`?

File: python/ray/projects/projects.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            return dir
+        prev, dir = dir, os.path.abspath(os.path.join(dir, os.pardir))
+    return None
+
+
+def validate_project(project_file):
```
Comment (simon-mo) [inline python/ray/projects/projects.py]: I think more validation logic need to be added:
- Make sure the autoscaler yaml file exists. 
- Make sure all the path exists. (Dockerfile, requirements, etc)

File: python/ray/projects/schema.json [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    "cluster": {
+      "description": "Path to a .yaml cluster configuration file (relative to the project root)",
+      "type": "string"
+    },
+    "environment": {
+      "description": "The environment that needs to be set up to run the project",
+      "type": "object",
+      "properties": {
+        "dockerfile": {
+          "description": "Path to a Dockerfile to set up an image the project can run in (relative to the project root)",
```
Comment (simon-mo) [inline python/ray/projects/schema.json]: This mean we will build the docker file and then start ray cluster inside it? In that case, do we also support `Docker Image` that we can directly pull from?

File: python/ray/projects/schema.json [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        "shell": {
+          "description": "A sequence of shell commands to run to set up the project environment",
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
+        }
+      }
+    },
+    "commands": {
```
Comment (simon-mo) [inline python/ray/projects/schema.json]: Need some example yaml in tests to see what commands means

File: python/ray/projects/projects.py (hunk 1 above)
Comment (pcmoritz) [reply python/ray/projects/projects.py]: This is a directory, which contains a project.yaml (and maybe also a cluster.yaml for the autoscaler).

File: python/ray/projects/projects.py (hunk 2 above)
Comment (pcmoritz) [reply python/ray/projects/projects.py]: Fixed!

File: python/ray/projects/schema.json (hunk 3 above)
Comment (pcmoritz) [reply python/ray/projects/schema.json]: I added a "dockerimage" field which can be used for this purpose

File: python/ray/projects/schema.json (hunk 4 above)
Comment (pcmoritz) [reply python/ray/projects/schema.json]: I scoped the commands down a little, added a description and added it to the project.yaml

File: python/ray/projects/projects.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    # Make sure the cluster yaml file exists
+    if "cluster" in project_definition:
+        cluster_file = os.path.join(project_root,
+                                    project_definition["cluster"])
+        assert os.path.exists(cluster_file)
+
+    if "environment" in project_definition:
+        env = project_definition["environment"]
+
+        if sum(["dockerfile" in env, "dockerimage" in env]) > 1:
```
Comment (simon-mo) [inline python/ray/projects/projects.py]:
```
if "dockerfile" in env and "dockerimage" in env:
```

File: python/ray/projects/projects.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            raise ValueError("Cannot specify both 'dockerfile' and "
+                             "'dockerimage' in environment.")
+
+        if "requirements" in env:
+            requirements_file = os.path.join(project_root, env["requirements"])
+            if not os.path.exists(requirements_file):
+                raise ValueError("'requirements' file in 'environment' does "
+                                 "not exist in {}".format(project_root))
+
+        if "dockerfile" in project_definition["environment"]:
```
Comment (simon-mo) [inline python/ray/projects/projects.py]:
```
if "dockerfile" in env
```

File: python/ray/projects/projects.py (hunk 5 above)
Comment (pcmoritz) [reply python/ray/projects/projects.py]: I worte it that way so if there is more in the future, it is easier to extend, what do you think?

File: python/ray/projects/projects.py (hunk 6 above)
Comment (pcmoritz) [reply python/ray/projects/projects.py]: Thanks, fixed!

**Review by simon-mo — APPROVED** [review]

File: python/ray/projects/projects.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
+import json
+import jsonschema
+import os
+import yaml
+
+
+def find_root(dir):
```
Comment (stephanie-wang) [inline python/ray/projects/projects.py]: Rename `dir -> directory` (`dir` is a builtin method in python).

File: python/ray/projects/projects.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        if os.path.isdir(os.path.join(dir, ".rayproject")):
+            return dir
+        prev, dir = dir, os.path.abspath(os.path.join(dir, os.pardir))
+    return None
+
+
+def validate_project_schema(project_definition):
+    """Validate a project file against the official ray project schema.
+
+    Raises an exception if the project file is not valid.
```
Comment (stephanie-wang) [inline python/ray/projects/projects.py]: In other parts of the Python code, we use a separate block for `Raises` and list the types of exceptions that can be raised ([example](https://github.com/ray-project/ray/blob/master/python/ray/worker.py#L261)).

File: python/ray/projects/projects.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    with open(os.path.join(dir, "schema.json")) as f:
+        schema = json.load(f)
+
+    jsonschema.validate(instance=project_definition, schema=schema)
+
+
+def check_project_definition(project_root, project_definition):
+    """Checks if the project definition is valid.
+
+    Raises an exception if the schema is not valid or if there are
```
Comment (stephanie-wang) [inline python/ray/projects/projects.py]: Same here, change to a `Raises` block.

File: python/ray/projects/projects.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        project_definition (dict): Project definition
+    """
+
+    validate_project_schema(project_definition)
+
+    # Make sure the cluster yaml file exists
+    if "cluster" in project_definition:
+        cluster_file = os.path.join(project_root,
+                                    project_definition["cluster"])
+        assert os.path.exists(cluster_file)
```
Comment (stephanie-wang) [inline python/ray/projects/projects.py]: Can we raise a `ValueError` here too for consistency?

File: python/ray/projects/projects.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            docker_file = os.path.join(project_root, env["dockerfile"])
+            if not os.path.exists(docker_file):
+                raise ValueError("'dockerfile' file in 'environment' does "
+                                 "not exist in {}".format(project_root))
+
+
+def load_project(current_dir):
+    """Finds .rayproject folder for current project, parse and validates it.
+
+    Args:
```
Comment (stephanie-wang) [inline python/ray/projects/projects.py]: Add a `Raises` block.

File: python/ray/scripts/scripts.py [hunk 12]
```diff
@@ -706,6 +706,17 @@ def get_worker_ips(cluster_config_file, cluster_name):
...
 
+@cli.command()
+@click.argument("command", required=True, type=str)
+@click.option(
+    "--dry",
+    is_flag=True,
+    default=False,
+    help="Print actions instead of running them.")
+def session(command, dry):
+    ray.projects.load_project(os.getcwd())
```
Comment (stephanie-wang) [inline python/ray/scripts/scripts.py]: I'm not sure if I understand this command...this is eventually how a session should be started, right? What is it supposed to do right now?

File: python/ray/projects/schema.json [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+          },
+          "command": {
+            "description": "Shell command to run on the cluster",
+            "type": "string"
+          }
+        }
+      }
+    }
+  },
+  "required": ["name"]
```
Comment (stephanie-wang) [inline python/ray/projects/schema.json]: Shouldn't there be more required fields? I could see requiring at least `repo` and `cluster`.

File: python/ray/projects/projects.py (hunk 7 above)
Comment (pcmoritz) [reply python/ray/projects/projects.py]: fixed

File: python/ray/projects/projects.py (hunk 8 above)
Comment (pcmoritz) [reply python/ray/projects/projects.py]: fixed

File: python/ray/projects/projects.py (hunk 9 above)
Comment (pcmoritz) [reply python/ray/projects/projects.py]: fixed

File: python/ray/projects/projects.py (hunk 10 above)
Comment (pcmoritz) [reply python/ray/projects/projects.py]: fixed

File: python/ray/projects/projects.py (hunk 11 above)
Comment (pcmoritz) [reply python/ray/projects/projects.py]: fixed

File: python/ray/scripts/scripts.py (hunk 12 above)
Comment (pcmoritz) [reply python/ray/scripts/scripts.py]: Right now it just validates the project file. @simon-mo will start working on the implementation that starts the session soon.

File: python/ray/projects/schema.json (hunk 13 above)
Comment (pcmoritz) [reply python/ray/projects/schema.json]: I added `cluster` for the requirement. I don't think `repo` is required, a project can just be a directory on disk without any associated repos.

**Review by stephanie-wang — APPROVED** [review]

**Pull request merged**